pkg/patterns/declarative/metrics.go

```diff
@@ -0,0 +1,600 @@
+}
+
+var _ workqueue.RateLimitingInterface = dummyQueue{}
+
+type dummyQueue struct{}
+
+func (dummyQueue) Add(item interface{})                              {}
+func (dummyQueue) Len() int                                          { return 0 }
+func (dummyQueue) Get() (item interface{}, shutdown bool)            { return struct{}{}, false }
+func (dummyQueue) Done(item interface{})                             {}
+func (dummyQueue) ShutDown()                                         {}
+func (dummyQueue) ShuttingDown() bool                                { return false }
+func (dummyQueue) AddAfter(item interface{}, duration time.Duration) {}
+func (dummyQueue) AddRateLimited(item interface{})                   {}
+func (dummyQueue) Forget(item interface{})                           {}
+func (dummyQueue) NumRequeues(item interface{}) int                  { return 0 }
+
+// Namespace & list of Name pairs
+type nsnPairs map[string][]string
+
+// ids holds namespace/name pairs and is indexed as ids[namespace][name]
+type items struct {
+	mu  sync.Mutex
+	ids map[string]bol
+}
+
+func (i *items) insert(namespace string, names ...string) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	if _, ok := i.ids[namespace]; !ok {
+		i.ids[namespace] = bol{}
+	}
+	i.ids[namespace].Insert(names...)
+
+}
+
+func (i *items) has(namespace, name string) bool {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	_, ok := i.ids[namespace]
+	return ok && i.ids[namespace].Has(name)
+}
+
+// Only used by another *items methods
+func (i *items) internalHas(namespace, name string) bool {
+	_, ok := i.ids[namespace]
+	return ok && i.ids[namespace].Has(name)
+}
+
+func (i *items) hasPairs(nsnp nsnPairs) bool {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	for ns, nlist := range nsnp {
+		for _, n := range nlist {
+			if !i.internalHas(ns, n) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (i *items) inc() {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	for ns := range i.ids {
+		i.ids[ns].Inc()
+	}
+}
+
+func (i *items) resetIfItHas(namespace, name string) bool {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	if _, has := i.ids[namespace]; has {
+		return i.ids[namespace].ResetIfItHas(name)
+	}
+	return false
+}
+
+func (i *items) markDeletedIfItHas(namespace, name string) bool {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	if _, has := i.ids[namespace]; has {
+		return i.ids[namespace].MarkDeletedIfItHas(name)
+	}
+	return false
+}
+
+func (i *items) deleteIfNeeded(metricsDuration int) (b bool, p nsnPairs) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	p = make(map[string][]string)
+	for ns := range i.ids {
+		for n := range i.ids[ns] {
+			if i.ids[ns].DeleteIfNeeded(n, metricsDuration) {
+				b = true
+				if _, ok := p[ns]; !ok {
+					p[ns] = []string{}
+				}
+				p[ns] = append(p[ns], n)
+			}
+		}
+		if len(i.ids[ns]) == 0 {
+			delete(i.ids, ns)
+		}
+	}
+
+	return b, p
+}
+
+func newItems() *items {
+	it := &items{}
+	it.ids = make(map[string]bol)
+	return it
+}
+
+type bol map[string]*counter
```

Comment:
Nit: I'm not sure what bol is short for, the full name would be clearer

Reply:
First I use `bol` as abbreviation of `bookOfLife`. However I'll use `record` instead of `bol`.